lib/bakeAmbientOcclusion.js

```diff
@@ -43,20 +48,323 @@ function bakeAmbientOcclusion(gltf, options) {
+        material.technique = cloneJsonAsNeeded(material.technique, parameters.techniqueCloneIDs, techniques);
+
+        var technique = techniques[material.technique];
+        technique.program = cloneJsonAsNeeded(technique.program, parameters.programCloneIDs, programs);
+
+        var program = programs[technique.program];
+        program.vertexShader = cloneJsonAsNeeded(program.vertexShader, parameters.vertexShaderCloneIDs, shaders);
+        program.fragmentShader = cloneJsonAsNeeded(program.fragmentShader, parameters.fragmentShaderCloneIDs, shaders);
+    }
+}
+
+function bakeToVertices(gltf, scene, raytracerScene, options) {
+
+    // Add the new vertex data to the buffers along with bufferViews and accessors
+    addVertexData(gltf, scene, raytracerScene);
 
-function bakeToDiffuse(gltf, scene, resolution, raytracerScene) {
+    // Build lists of shaders, techniques, and primitives that need to be edited.
+    // Basically, edit anything used by a primitive that has AO data lying around.
+    // If anything is used by a primitive that doesn't have per-vertex AO, clone it first.
+    var parameters = {
+        aoBufferByPrimitive: raytracerScene.aoBufferByPrimitive,
+        materials: gltf.materials,
+        techniques: gltf.techniques,
+        programs: gltf.programs,
+        shaders: gltf.shaders,
+        // Use the keys to keep track of what needs editing.
+        // If an item gets cloned, store the clone's name as the value.
+        materialCloneIDs: {},
+        techniqueCloneIDs: {},
+        programCloneIDs: {},
+        vertexShaderCloneIDs: {},
+        fragmentShaderCloneIDs: {}
+    };
+
+    NodeHelpers.forEachPrimitiveInScene(gltf, scene, checkShadingChain, parameters);
+
+    // Clone materials, techniques, programs, and shaders as needed by primitives in other scenes.
+    for (var otherSceneID in gltf.scenes) {
+        if (gltf.scenes.hasOwnProperty(otherSceneID)) {
+            var otherScene = gltf.scenes[otherSceneID];
+            NodeHelpers.forEachPrimitiveInScene(gltf, otherScene, cloneShadingChain, parameters);
+        }
+    }
+
+    // Edit the shaders.
+    // This can mean tracking down the unlit diffuse color input from the technique.
+    addAoToShaders(gltf, Object.keys(parameters.techniqueCloneIDs), options);
+
+    // Edit the programs: add the ao attribute
+    for (var programID in parameters.programCloneIDs) {
+        if (parameters.programCloneIDs.hasOwnProperty(programID)) {
+            gltf.programs[programID].attributes.push('a_ambientOcclusion');
+        }
+    }
+
+    // Edit the techniques: add ao to attributes, add ao to parameters
+    for (var techniqueID in parameters.techniqueCloneIDs) {
+        if (parameters.techniqueCloneIDs.hasOwnProperty(techniqueID)) {
+            var technique = gltf.techniques[techniqueID];
+            technique.attributes.a_ambientOcclusion = 'vertex_ao';
+            technique.parameters.vertex_ao = {
+                semantic: 'VERTEX_AO',
```

Comment:
As discussed offline: it seems like we need the semantic, so we'll use `_OCCLUSION` and document it in case of conflicts with user provided custom semantics.